include/NovelRT/Maths/QuadTree.h

```diff
@@ -0,0 +1,135 @@
+      return _bounds;
+    }
+
+    std::shared_ptr<TQuadTreePoint> getPoint(uint32_t index) const noexcept {
+      return (index < _pointCount) ? _points[index] : nullptr;
+    }
+
+    uint32_t getPointCount() const noexcept {
+      return _pointCount;
+    }
+
+    QuadTree* getTopLeft() const noexcept {
+      return _children[TOP_LEFT].get();
+    }
+
+    QuadTree* getTopRight() const noexcept {
+      return _children[TOP_RIGHT].get();
+    }
+
+    QuadTree* getBottomLeft() const noexcept {
+      return _children[BOTTOM_LEFT].get();
+    }
+
+    QuadTree* getBottomRight() const noexcept {
+      return _children[BOTTOM_RIGHT].get();
+    }
+
+    bool tryInsert(std::shared_ptr<TQuadTreePoint> point) noexcept {
+      if (point == nullptr || !getBounds().pointIsWithinBounds(point->getPosition())) return false;
+
+      if (_children[TOP_LEFT] == nullptr) {
+        if (_pointCount < POINT_CAPACITY) {
+          _points[_pointCount++] = point;
+          return true;
+        }
+        subdivideTree();
+      }
+
+      auto result = _children[TOP_LEFT]->tryInsert(point) ||
+                    _children[TOP_RIGHT]->tryInsert(point) ||
+                    _children[BOTTOM_LEFT]->tryInsert(point) ||
+                    _children[BOTTOM_RIGHT]->tryInsert(point);
+
+      assert(result);
+      return result;
+    }
+
+    void subdivideTree() noexcept {
+      const GeoVector<float> TOP_LEFT_SCALE = GeoVector<float>(-0.5, +0.5);
+      const GeoVector<float> TOP_RIGHT_SCALE = GeoVector<float>(+0.5, +0.5);
+      const GeoVector<float> BOTTOM_LEFT_SCALE = GeoVector<float>(-0.5, -0.5);
+      const GeoVector<float> BOTTOM_RIGHT_SCALE = GeoVector<float>(+0.5, -0.5);
+
+      GeoVector size = getBounds().getSize() / 2;
+      GeoVector position = getBounds().getPosition();
+
+      _children[TOP_LEFT] = std::make_unique<QuadTree>(GeoBounds(position + (size * TOP_LEFT_SCALE), size, 0));
+      _children[TOP_RIGHT] = std::make_unique<QuadTree>(GeoBounds(position + (size * TOP_RIGHT_SCALE), size, 0));
+      _children[BOTTOM_LEFT] = std::make_unique<QuadTree>(GeoBounds(position + (size * BOTTOM_LEFT_SCALE), size, 0));
+      _children[BOTTOM_RIGHT] = std::make_unique<QuadTree>(GeoBounds(position + (size * BOTTOM_RIGHT_SCALE), size, 0));
+
+      for (uint32_t index = 0; index < getPointCount(); index++) {
+        auto point = getPoint(index);
+        auto result = tryInsert(point);
+        assert(result);
+      }
+
+      _pointCount = 0;
+    }
+
+    void getIntersectingPoints(const GeoBounds& bounds, std::vector<std::shared_ptr<TQuadTreePoint>>& intersectingPoints) {
+      if (getBounds().intersectsWith(bounds)) {
+        return;
+      }
+
+      if (_children[TOP_LEFT] == nullptr) {
+        for (uint32_t index = 0; index < getPointCount(); index++) {
+          auto point = getPoint(index);
+          if (bounds.pointIsWithinBounds(point->getPosition())) {
+            intersectingPoints.push_back(point);
```

Comment:
would `emplace_back` not be better here?

Reply:
Now that this is `std::shared_ptr`, yes. Will fix. There would have been no difference when it was just `T*` since that is a single write either way.

Reply:
Ah I see, awesome.